apt/method.go

```diff
@@ -0,0 +1,225 @@
+		if err != nil {
+			// TODO: log?
+			continue
+		}
+		switch msg.code {
+		case 600:
+			m.handleAcquire(msg)
+		case 601:
+			m.handleConfigure(msg)
+		default:
+			m.writer.Fail("Unsupported API")
+		}
+	}
+}
+
+func (m *AptMethod) getClient() error {
+	if m.client != nil {
+		return nil
+	}
+
+	var ts oauth2.TokenSource
+	ctx := context.Background()
+	switch {
+	case m.config.serviceAccountJSON != "":
+		json, err := ioutil.ReadFile(m.config.serviceAccountJSON)
+		if err == nil {
+			creds, ierr := google.CredentialsFromJSON(ctx, json)
+			if ierr == nil {
+				ts = creds.TokenSource
+			}
+		}
+	case m.config.serviceAccountEmail != "":
+		ts = google.ComputeTokenSource(m.config.serviceAccountEmail)
+	default:
+		creds, err := google.FindDefaultCredentials(ctx)
+		if err == nil {
+			ts = creds.TokenSource
+		}
+	}
+	if ts == nil {
+		return errors.New("Failed to obtain token source")
+	}
+	m.client = oauth2.NewClient(ctx, ts)
+	return nil
+}
+
+func (r realDownloader) download(body io.ReadCloser, filename string) (string, error) {
+	defer body.Close()
+	data, err := ioutil.ReadAll(body)
+	if err != nil {
+		return "", err
+	}
+	file, err := os.Create(filename)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	_, err = file.Write(data)
+	return fmt.Sprintf("%x", md5.Sum(data)), err
+}
+
+// TODO: testing. this validates the message (could be separated), optionally adds a header, makes a GET request, and sends appropriate 2xx messages to apt.
+// it also returns an error that we don't care about..
+func (m *AptMethod) handleAcquire(msg *AptMessage) error {
+	// === VALIDATE MESSAGE ===
+	uri := msg.Get("URI")
+	if uri == "" {
+		err := errors.New("No URI provided")
+		m.writer.Fail(err.Error())
+		return err
+	}
+	filename := msg.Get("Filename")
+	if filename == "" {
+		err := errors.New("No filename provided")
+		m.writer.FailURI(uri, err.Error())
+		return err
+	}
+	ifModifiedSince := msg.Get("Last-Modified")
+	// === END VALIDATE MESSAGE ===
+
+	// === GET CLIENT ===
+	if err := m.getClient(); err != nil {
+		m.writer.FailURI(uri, err.Error())
+		return err
+	}
+	// === END GET CLIENT ===
+
+	// === MAKE REQUEST ===
+	realuri := strings.Replace(uri, "ar+https", "https", 1)
+	req, err := http.NewRequest("GET", realuri, nil)
+	if err != nil {
+		return err
+	}
+	if ifModifiedSince != "" {
+		// TODO: validate this string is in RFC1123Z format.
+		req.Header.Add("If-Modified-Since", ifModifiedSince)
+	}
+	resp, err := m.client.Do(req)
+	if err != nil {
+		m.writer.FailURI(uri, err.Error())
+		return err
+	}
+	// === END MAKE REQUEST ===
+
+	// === HANDLE RESPONSE ===
+	size := resp.Header.Get("Content-Length")
+	lastModified := resp.Header.Get("Last-Modified")
+	switch resp.StatusCode {
+	case 200:
+		// It's weird to send URI Start after we've already contacted
+		// the server, but we need to know the size.
+		m.writer.URIStart(uri, size, lastModified)
+		if m.dl == nil {
+			m.dl = realDownloader{}
+		}
```

Comment:
can downloader be initialized when message is created or is it too expensive so lazy init like here is needed?

Reply:
downloader literally only exists to let me unit test this function, which would otherwise try to write a body to a file. if there is a better way to accomplish this i would appreciate the suggestion :)

Reply:
moved to NewAptMethod